src/java/org/apache/cassandra/io/util/CompressedChunkReader.java

```diff
@@ -117,8 +141,23 @@ public void readChunk(long position, ByteBuffer uncompressed)
                 {
                     ByteBuffer compressed = bufferHolder.getBuffer(length);
 
-                    if (channel.read(compressed, chunk.offset) != length)
-                        throw new CorruptBlockException(channel.filePath(), chunk);
+                    if (readAheadBuffer != null && readAheadBuffer.hasBuffer())
+                    {
+                        int copied = 0;
+                        while (copied < length) {
+                            readAheadBuffer.fill(chunk.offset + copied);
```

Comment:
this feels dangerous as you don't track the offsets... you are asked for `position`, and you assume the buffer is read from and mutated... this feels like it would be very easy to cause data corruption.

Reply:
@dcapwell open to suggested improvements. not totally sure I understand what you mean by "track" the offsets. i largely based it on the existing code that reads directly from the `FileChannel` and how the `compressed` buffer is treated in the original code to minimize changes.